instant-distance/src/contiguous.rs

```diff
@@ -0,0 +1,240 @@
+use std::marker::PhantomData;
+
+use rand::{rngs::SmallRng, Rng, SeedableRng};
+#[cfg(feature = "serde")]
+use serde::{Deserialize, Serialize};
+
+use crate::{
+    types::{LayerId, Meta, INVALID},
+    Builder, Point, PointId,
+};
+
+pub trait PointDataSource: Sync {
+    fn decompose(&self) -> Vec<f32>;
+    fn stride() -> usize;
+}
+
+pub struct PointIter<'a> {
+    values: &'a [f32],
+    order: &'a [usize],
+    stride: usize,
+    index: usize,
+}
+
+impl<'a> Iterator for PointIter<'a> {
+    type Item = PointRef<'a>;
+
+    fn next(&mut self) -> Option<Self::Item> {
+        if self.index < self.order.len() {
+            let i = self.order[self.index] * self.stride;
+            self.index += 1;
+            Some(PointRef(&self.values[i..i + self.stride]))
+        } else {
+            None
+        }
+    }
+}
+
+pub trait Storage<P: PointDataSource> {
+    fn iter(&self) -> PointIter;
+    fn get(&self, index: usize) -> Option<PointRef>;
+    fn len(&self) -> usize;
+    fn is_empty(&self) -> bool;
+}
+
+pub struct PointRef<'a>(pub &'a [f32]);
+
+impl<'a> PointRef<'a> {
+    pub fn from_data(values: &'a [f32]) -> Self {
+        Self(values)
+    }
+}
+
+#[cfg_attr(feature = "serde", derive(Deserialize, Serialize))]
+#[derive(Default)]
+pub struct ContiguousStorage<T: PointDataSource> {
+    pub values: Vec<f32>,
+    pub order: Vec<usize>,
```

Comment:
What is `order` for?

Reply:
In the original code from the `neighbors` branch, nodes are reordered and assigned layers. This is then stored. I wanted to preserve the reordering, but not mutate the underlying store, which I believe will be needed when we get to mmap-ing or working with point data from something like lmdb. This introduces an indirection, but in my testing, hasn't negatively impacted performance. This is another design consideration I'm happy to hear your thoughts on.

Reply:
I think we have to randomize the order points on construction, and then make sure to avoid further reordering after construction. The way ordering works now is that we give the caller a `Vec<PointId>` after construction which avoids the `Hnsw` having to keep track of the original order, IIRC.

Reply:
Ok `order` is gone now and we keep the backing store shuffled as before.